### PR TITLE
add option for different plasticity for different phases

### PR DESCRIPTION
**Pull request by elodie-kendall**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*
Following problems of plastic yielding in the lower mantle and therefore large viscosity contrasts, I added an option for different plasticity for different phases.

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Review by naliboff — COMMENTED** [review]
@elodie-kendall - Thank you for tackling this! I had an initial read through and the implementation had quite a few more changes than I had expected. This may end up being the correct approach, but I think we should speak with @bobmyhill and @anne-glerum about the overall approach and ideas for simplifying. I need to go through the current code structure again and also think about how these changes will affect the composite rheology option @bobmyhill and I have been working on.

**Comment by naliboff** [comment]
@elodie-kendall - I had a quick look through the code again and I think I understand why so much may changes to interfaces and variable definitions took place:

- `MaterialModel::MaterialUtilities::phase_average_value` is passed values for a single compositional field.
- In most cases the functions in the drucker_prager rheology module are passed values for a single composition, but in the function `compute_strain_rate_and_derivative` the entire set of drucker prager values (`drucker_prager_parameters`) are inputs along with the composition.
- As a result, you now compute the drucker prager parameters within `compute_strain_rate_and_derivative` with the new function `compute_drucker_prager_parameters`

Is this a reasonable summary? 

This may be the approach we need to take, but the test using `compute_strain_rate_and_derivative` is now failing due to an error when parsing the parameters from drucker prager. This test involves compiling a plugin for the composite rheology, so not surprising.

Let's discuss this tomorrow with @bobmyhill and @anne-glerum. Ideally, we would change as little as possible to the current variable definitions and interfaces, but this is something to discuss as a group. 

John

File: source/material_model/rheology/drucker_prager.cc
Comment (anne-glerum) [inline source/material_model/rheology/drucker_prager.cc:237]: rm empty line

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (anne-glerum) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:8]: Several nonlinear iterations would be better. Maybe set to 10? Or is that really slow?

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (anne-glerum) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:1]: Can you add one other sentence that says what this test specifically tests? I.e. different plasticity parameters for different phases.

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (anne-glerum) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:78]: A friction angle of 0.1 doesn't seem like a value that we would use. What about 20 degrees?

Comment (anne-glerum) [reply tests/visco_plastic_differentplasticitydifferentphases.prm:78]: Also, these phase values are ordered according to increasing depth right? Maybe switch around the values then, so that the crust-like values (20 degrees, 10 MPa) come first.

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (anne-glerum) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:97]: What about `named additional outputs` so that the cohesion and friction angle are also outputted?

File: include/aspect/material_model/rheology/drucker_prager.h
Comment (anne-glerum) [inline include/aspect/material_model/rheology/drucker_prager.h:46]: Since you now only return one value, please update the comment (it's no longer a list of angles) and change the name of the parameter to `angle_internal_friction` or `internal_friction_angle`

Comment (anne-glerum) [reply include/aspect/material_model/rheology/drucker_prager.h:46]: same for cohesions

File: include/aspect/material_model/rheology/drucker_prager.h
Comment (anne-glerum) [inline include/aspect/material_model/rheology/drucker_prager.h:87]: one empty line between function definitions

File: include/aspect/material_model/rheology/drucker_prager.h
Comment (anne-glerum) [inline include/aspect/material_model/rheology/drucker_prager.h:102]: Please keep the empty line between function definitions.

File: include/aspect/material_model/rheology/drucker_prager.h
Comment (anne-glerum) [inline include/aspect/material_model/rheology/drucker_prager.h:119]: Why do you need the phase information here? Can you compute the DruckerPragerParameters before you call this function?

File: include/aspect/material_model/rheology/visco_plastic.h
Comment (anne-glerum) [inline include/aspect/material_model/rheology/visco_plastic.h:183]: Please undo the removal of the space before the asterix. It should look like this:
```
/**
 *
 */
```

File: source/material_model/rheology/composite_visco_plastic.cc
Comment (anne-glerum) [inline source/material_model/rheology/composite_visco_plastic.cc:200]: In line 170 above, drucker_prager_parameters is already computed. Could you use that here instead of passing the phase_function_values and n_phases_per_composition?

**Review by anne-glerum — CHANGES_REQUESTED** [review]
Hi Elodie! Thanks for this PR. I put in some small comments about empty lines and spaced, and comments about how I think we can avoid changing most of the function interfaces. It could be that I'm missing something that makes this not possible, so let's discuss later today with @naliboff  and @bobmyhill as John suggested.

File: tests/visco_plastic_differentplasticitydifferentphases/screen-output
Comment (bobmyhill) [inline tests/visco_plastic_differentplasticitydifferentphases/screen-output:76]: It's usually easiest to move and rename the file generated by ctest (the one with the suffix .notime), which doesn't have all the version and timing information. Or you can manually remove all lines beginning `--`, `|` and `+`.

**Review by bobmyhill — COMMENTED** [review]
I don't have much to add to @anne-glerum's review. I think the peierls_creep rheology model should serve as a good template for this PR.

File: include/aspect/material_model/rheology/visco_plastic.h
Comment (naliboff) [inline include/aspect/material_model/rheology/visco_plastic.h:179]: Combine this line and the previous one for readability?

File: include/aspect/material_model/rheology/drucker_prager.h
Comment (naliboff) [inline include/aspect/material_model/rheology/drucker_prager.h:112]: I think you want to keep this line?

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (naliboff) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:1]: I would consider renaming the file to something simpler like `visco_plastic_phases_plasticity` or `visco_plastic_phases_plastic_parameters`?

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (naliboff) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:8]: You can see this value to 1 for the test

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (naliboff) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:9]: Set to the name of the test? (although this does not matter when the tester is run, parameter is ignored)

File: tests/visco_plastic_differentplasticitydifferentphases.prm
Comment (naliboff) [inline tests/visco_plastic_differentplasticitydifferentphases.prm:43]: Fine for now, but in a follow-up pull request this could be changed to 
```suggestion
  set List of model names = initial temperature
```
, which would allow deleting the subsection below

**Review by naliboff — COMMENTED** [review]
@elodie-kendall - Nice work! Overall the structure looks correct. I will take a look through later and add more detailed comments for formatting/style, but for now let's make sure all the tests are passing.

File: tests/visco_plastic_phases_plasticity.prm
Comment (naliboff) [inline tests/visco_plastic_phases_plasticity.prm:1]: Can you add some additional comments here describing what happens (e.g., how values change above and below phase transition) and what files this should be visible in (e.g., ascii visualization output)?

File: tests/visco_plastic_phases_plasticity.prm
Comment (naliboff) [inline tests/visco_plastic_phases_plasticity.prm:97]: Add the following line here to specify an ascii output format:
`set Output format            = gnuplot`
After re-running the test with this parameter set, then add the files in the solution folder so they are included in the test suite!

**Review by naliboff — COMMENTED** [review]
@elodie-kendall - It looks like you were updating and fixing the tests as I was reviewing and making comments 😄 

I think it would be good to out the visualization output (ascii format) to the test suite to ensure we are getting the correct cohesion and friction angles above and below the phase transition. See the comments for more details.

It looks like the composite visco plastic test also needs to be updated and I will look at that next. The other failing tests (Steinberger) will require a rebase and force push (likely discussed in morning rounds).

Getting close!

**Comment by naliboff** [comment]
@elodie-kendall - I just had a look at the tester outputs and indeed the one failing test (composite_viscous_outputs) is one that involves compiling additional code ([composite_viscous_outputs.cc](https://github.com/geodynamics/aspect/blob/master/tests/composite_viscous_outputs.cc)) to check whether the composite rheology models are producing the correct outputs.

There are a number of errors related to changes in this PR [(link)](https://jenkins.tjhei.info/blue/organizations/jenkins/aspect/detail/PR-4161/13/tests), but as expected most involve `composite_visco_plastic` or `drucker_prager`.

Can you go through that test and make the necessary adjustments? Let us know if you run into any issues or would like to discuss this directly.

In order to see if your changes work locally, after doing `make setup_tests` in the builder folder you can run `cmake -R composite_viscous_outputs`.

After that test is finished, I think this will be effectively ready to merge after one final pass through!

File: include/aspect/material_model/rheology/drucker_prager.h
Comment (naliboff) [reply include/aspect/material_model/rheology/drucker_prager.h:86]: @elodie-kendall - can you add a space between this line and the next one?

File: source/material_model/rheology/drucker_prager.cc
Comment (naliboff) [inline source/material_model/rheology/drucker_prager.cc:213]: rm empty line

File: source/material_model/rheology/drucker_prager.cc
Comment (naliboff) [reply source/material_model/rheology/drucker_prager.cc:231]: @elodie-kendall

File: tests/visco_plastic_phases_plasticity.prm
Comment (naliboff) [inline tests/visco_plastic_phases_plasticity.prm:1]: Maybe break this comment down into multiple lines?

**Review by naliboff — COMMENTED** [review]
@elodie-kendall - Fantastic, if the tester passes this is basically good to go! Thank you again for all your work on this! 

The final review comments from my side are just small formatting cleanup, but after that this can be merged.

**Review by naliboff — APPROVED** [review]
@elodie-kendall - the testers are happy 🍾 ! 

Can you squash the last two commits and have a commit message along the lines of "address comments"? Once that is done, ready to merge from my side. 

@anne-glerum @bobmyhill - Would you like a final look as well?

**Comment by elodie-kendall** [comment]
Thanks John. I squashed the commits, so it should be ready to be merge:)

File: source/material_model/rheology/drucker_prager.cc
Comment (anne-glerum) [inline source/material_model/rheology/drucker_prager.cc:131]: For consistency, can you use `p.max_yield_stress` here?

File: source/material_model/rheology/drucker_prager.cc
Comment (anne-glerum) [inline source/material_model/rheology/drucker_prager.cc:207]:
```suggestion
        // Retrieve the list of composition names
```

File: tests/visco_plastic_phases_plasticity.prm
Comment (anne-glerum) [inline tests/visco_plastic_phases_plasticity.prm:4]:
```suggestion
# This is visible in the visualisation output.
```

File: include/aspect/material_model/rheology/drucker_prager.h
Comment (anne-glerum) [inline include/aspect/material_model/rheology/drucker_prager.h:85]:
```suggestion
                            const std::unique_ptr<std::vector<unsigned int>> &expected_n_phases_per_composition =
                              nullptr);
```

**Review by anne-glerum — APPROVED** [review]
Hi Elodie, wonderful! Sorry, but there are just a few tiny things left. Once they're fixed this can be merged. These edits shouldn't change any output, so it should be quick! Thanks for making this work!

**Pull request merged**
